h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -0,0 +1,720 @@
+                    }
+                }
+
+                for (int j = 0; j < di._nums; ++j) {
+                    float val = (float) vecs[di._cats + j].at(i);
+                    if (!Float.isNaN(val) && val != 0) {
+                        data[currentRow][currentCol] = val;
+                        colIndex[currentRow][currentCol++] = di._catOffsets[di._catOffsets.length - 1] + j;
+                        nz++;
+                    }
+                }
+                if (nz == nzstart) {
+                    // for the corner case where there are no categorical values, and all numerical values are 0, we need to
+                    // assign a 0 value to any one column to have a consistent number of rows between the predictors and the special vecs (weight/response/etc.)
+                    data[currentRow][currentCol] = 0;
+                    colIndex[currentRow][currentCol++] = 0;
+                    nz++;
+                }
+                rowHeaders[0][++actualRows] = nz;
+
+                rwRow = setResponseAndWeight(w, resp, weights, respVec, rwRow, i);
+            }
+        }
+
+        data[data.length - 1] = Arrays.copyOf(data[data.length - 1], nz % ARRAY_MAX);
+        colIndex[colIndex.length - 1] = Arrays.copyOf(colIndex[colIndex.length - 1], nz % ARRAY_MAX);
+
+        rowHeaders[0] = Arrays.copyOf(rowHeaders[rowHeaders.length - 1], actualRows + 1);
+        return actualRows;
+    }
+
+    private static int initalizeFromChunks(Chunk[] chunks, int weight, DataInfo di, int actualRows, long[][] rowHeaders, float[][] data, int[][] colIndex, int respIdx, float[] resp, float[] weights) {
+        int nz = 0;
+        int currentRow = 0;
+        int currentCol = 0;
+        int rwRow = 0;
+
+        for (int i = 0; i < chunks[0].len(); i++) {
+            if (weight != -1 && chunks[weight].atd(i) == 0) continue;
+            int nzstart = nz;
+
+            enlargeTables(data, colIndex, di._cats + di._nums, currentRow, currentCol);
+
+            for (int j = 0; j < di._cats; ++j) {
+                if (!chunks[j].isNA(i)) {
+                    data[currentRow][currentCol] = 1; //one-hot encoding
+                    colIndex[currentRow][currentCol++] = di.getCategoricalId(j, chunks[j].at8(i));
+                    nz++;
+                } else {
+                    // NA == 0 for sparse -> no need to fill
+//            data[nz] = 1; //one-hot encoding
+//            colIndex[nz] = di.getCategoricalId(j, Double.NaN); //Fill NA bucket
+//            nz++;
+                }
+            }
+            for (int j = 0; j < di._nums; ++j) {
+                float val = (float) chunks[di._cats + j].atd(i);
+                if (!Float.isNaN(val) && val != 0) {
+                    data[currentRow][currentCol] = val;
+                    colIndex[currentRow][currentCol++] = di._catOffsets[di._catOffsets.length - 1] + j;
+                    nz++;
+                }
+            }
+            if (nz == nzstart) {
```

Comment:
This part of code seems to be almost the same as the one starting with the same comment above